server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -384,7 +386,11 @@ public SearchService(
     }
 
     private void setEnableSearchWorkerThreads(boolean enableSearchWorkerThreads) {
-        this.enableSearchWorkerThreads = enableSearchWorkerThreads;
+        if (enableSearchWorkerThreads) {
+            searchExecutor = threadPool.executor(Names.SEARCH);
```

Comment:
> as soon as we have queuing past a certain pool size we stop parallelising

Good point. In designing this mechanism, some further cases to consider/prevent would be if:
- slice tasks are added to a full queue, which rejects them
- slice tasks are added to a full-ish queue, which then has latency implications
- slice tasks are added to a full-ish queue, causing their memory to be retained until all the preceding search requests are moved through the queue.

Reply:
> IMO: what should happen here is that as soon as we have queuing past a certain pool size we stop parallelising

I remember discussing this with Luca and maybe a couple other people in the past, this sounds like an easy/quick win to me. Search concurrency in Lucene is a way to trade throughput for latency (things like dynamic pruning are not perfectly parallelizable). As soon as your seach queue reaches a certain size, your latency is likely driven by the time you need to wait for your items before your query in the queue to process rather than by the time it takes your very query to process, so parallelizing the query likely barely helps. At the same time, no longer parallelizing requests would improve throughput, and thus help empty the queue more quickly.

Reply:
Agreed, it feels natural to disable inter-segment concurrency as a whole if we see a queue, and possibly even if we are close to queuing. That's because using a single pool would otherwise causes earlier 429s otherwise, depending on how many shards and segments a search is targeting.

Lucene does not always parallelize based on slices (think knn query rewrite for instance), so adjusting the number of slices gradually would not achieve the same goal. We can instead simply stop setting the executor to the searcher when we detect that parallelizing searches won't help. Besides how to avoid 429s like I said above, I think that a more adaptive behaviour would help get the best performance: when there's room, parallelize and use the available resources. optimize for latency. When there's less room, stop parallelizing and optimize for throughput which in turn helps reducing queues and emptying them more quickly.

Reply:
++ to not setting an executor if we have queuing, that gives us a low-contention and simple to implement fix here :) let me work that in real quick

Reply:
Should it be done in the main repo rather than on this branch since it feels generally desirable, not only if we're combining executors? (Sorry for the noise if this is what you had in mind already.)

Reply:
The problem with doing it in the main repo is that as soon as search_worker is fully busy, passing `null` means adding hot work on the current pool (likely "search"), thus upping concurrency levels a lot. That seems like it could way oversubscribe CPUs by potentially doubling the parallelism?

Reply:
Sorry I got confused, you are right, let's avoid double parallelism. The best we could do would be to create a single slice, which doesn't address all the challenges as Luca pointed out.

Reply:
How about? https://github.com/elastic/elasticsearch/pull/111099/commits/4fee515fcd1abf029a51fb7f6f0427266d8281fb 

I somewhat hated the fact that we need to cast/type-check the executor, so I figured I'd reuse a place where we already do it anyway :)